src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -90,25 +90,81 @@ public Dictionary(int capacity, IEqualityComparer<TKey>? comparer)
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.dictionary);
             }
 
+            if (dictionary.Count == 0)
+            {
+                // Nothing to copy, all done
+                return;
+            }
+
+            Debug.Assert(_buckets is not null);
+            Debug.Assert(_entries is not null);
+
             // It is likely that the passed-in dictionary is Dictionary<TKey,TValue>. When this is the case,
             // avoid the enumerator allocation and overhead by looping through the entries array directly.
             // We only do this when dictionary is Dictionary<TKey,TValue> and not a subclass, to maintain
             // back-compat with subclasses that may have overridden the enumerator behavior.
             if (dictionary.GetType() == typeof(Dictionary<TKey, TValue>))
             {
                 Dictionary<TKey, TValue> d = (Dictionary<TKey, TValue>)dictionary;
+
+                Debug.Assert(d._buckets is not null);
+                Debug.Assert(d._entries is not null);
+
+                Entry[] oldEntries = d._entries;
+                // If comparers are the same, we can copy dictionary wholesale without rehashing.
+                if (d._comparer == _comparer)
+                {
+                    Entry[] newEntries = _entries;
+                    // If dictionary is the same size for the number of items, copy the arrays directly.
+                    int length = oldEntries.Length;
+                    if (newEntries.Length == length)
+                    {
+                        Array.Copy(oldEntries, 0, newEntries, 0, length);
```

Comment:
It is not necessary to pass in zero start indices here. The overload with fewer arguments is faster now.